pkgpanda/actions.py

```diff
@@ -0,0 +1,293 @@
+
+    swap(install, repository, 'foo--version') will replace the active 'foo'
+    package with 'foo--version'.
+
+    install: pkgpanda.Install
+    repository: pkgpanda.Repository
+    package_id: package ID to activate
+    systemd: start/stop systemd services (default: True)
+    block_systemd: if systemd, block waiting for systemd services to come up (default: False)
+
+    """
+    active = install.get_active()
+    # TODO(cmaloney): I guarantee there is a better way to write this and
+    # I've written the same logic before...
+    packages_by_name = dict()
+    for id_str in active:
+        pkg_id = PackageId(id_str)
+        packages_by_name[pkg_id.name] = pkg_id
+
+    new_id = PackageId(package_id)
+    if new_id.name not in packages_by_name:
+        print("ERROR: No package with name {} currently active to swap with.".format(new_id.name))
+
+    packages_by_name[new_id.name] = new_id
+    new_active = list(map(str, packages_by_name.values()))
+    # Activate with the new package name
+    activate_packages(install, repository, new_active, systemd, block_systemd)
+
+
+def fetch_packages(repository, repository_url, package_ids, work_dir=None):
+    """Fetch package_ids from repository_url into repository.
+
+    repository: pkgpanda.Repository
+    repository_url: URL for remote package repository
+    package_ids: sequence of package IDs to fetch
+    work_dir: location for temporary files (default: current working dir)
+
+    """
+    assert isinstance(package_ids, collections.Sequence)
+    work_dir = work_dir or os.getcwd()
+
+    def fetcher(id_, target):
+        return requests_fetcher(repository_url, id_, target, work_dir)
+
+    for pkg_id in package_ids:
+        # TODO(cmaloney): Make this not use escape sequences when not at a
+        # `real` terminal.
+        sys.stdout.write("\rFetching: {0}".format(pkg_id))
+        sys.stdout.flush()
+        try:
+            repository.add(fetcher, pkg_id)
+        except FetchError as ex:
+            print("\nUnable to fetch package {0}: {1}".format(pkg_id, ex))
+            sys.exit(1)
+        sys.stdout.write("\rFetched: {0}\n".format(pkg_id))
+        sys.stdout.flush()
+
+
+def add_package_file(repository, package_filename):
+    """Add a package to the repository from a file.
+
+    repository: pkgpanda.Repository
+    package_filename: location of the package file
+
+    """
+    filename_suffix = '.tar.xz'
+    # Extract Package Id (Filename must be path/{pkg-id}.tar.xz).
+    name = os.path.basename(package_filename)
+
+    if not name.endswith(filename_suffix):
+        print("ERROR: Can only add package tarballs which have names "
+              "like {{pkg-id}}{}".format(filename_suffix))
+
+    pkg_id = name[:-len(filename_suffix)]
+
+    # Validate the package id
+    PackageId(pkg_id)
+
+    def fetch(_, target):
+        extract_tarball(package_filename, target)
+
+    repository.add(fetch, pkg_id)
+
+
+def remove_packages(install, repository, package_ids):
```

Comment:
Same thing applies for this as fetch. This intentionally only took a single package ID string.